rs/state_manager/src/lib.rs

```diff
@@ -3493,6 +3493,16 @@ impl StateManager for StateManagerImpl {
             Self::compute_certification_metadata(&self.metrics, &self.log, &state)
                 .unwrap_or_else(|err| fatal!(self.log, "Failed to compute hash tree: {:?}", err));
 
+        // This step is expensive, so we do it before the write lock for `states`.
+        let next_tip = {
+            let _timer = self
+                .metrics
+                .checkpoint_op_duration
+                .with_label_values(&["copy_state"])
+                .start_timer();
+            Some((height, state.deref().clone()))
+        };
+
         let mut states = self.states.write();
```

Comment:
Definitely not for this PR, but I'm wondering whether it wouldn't make sense for the code below this line to be in its own, separate function, documented to hold this lock. Or at least in an inline block with a comment.

I remember that Google's C++ style guide required specific annotations on functions (with meanings such as "acquires lock X" or "requires lock Y to be held"). These would also be validated by some tool (against e.g. deadlocks, if two locks could be acquired in different order on different code paths), but we don't need to go so far, since we only have a handful of (meaningful) mutexes throughout the codebase. It would be worth it simply as documentation.

Reply:
I think this could be a good thing to do. We should then also get rid of the awkward `release_lock_and_persist_metadata` function, which solves a similar issue of releasing the lock as early as possible. 

But for the context of this PR I'm gonna ignore the comment.